src/backends/interflop-mca/interflop_mca.c

```diff
@@ -181,22 +202,47 @@ static void _set_mca_precision_binary64(const int precision) {
  * perturbations used for MCA
  ***************************************************************/
 
-/* random generator internal state */
+/* random number generator internal state */
 static tinymt64_t random_state;
 
-static double _mca_rand(void) {
+/* random number generator internal state for simple generators */
+// __declspec(thread) static unsigned int random_state_simple;
+// __declspec(thread) static bool random_state_simple_valid;
+static __thread unsigned int random_state_simple;
+static __thread bool random_state_simple_valid = false;
+
+static pthread_mutex_t global_tid_lock;
+static int global_tid;
+
+static double _mca_rand(void *context) {
   /* Returns a random double in the (0,1) open interval */
-  return tinymt64_generate_doubleOO(&random_state);
+  if (MCALIB_RNG_MODE == mca_rng_mode_mt)
+    return tinymt64_generate_doubleOO(&random_state);
+  else {
+    if (random_state_simple_valid == false) {
```

Comment:
You could maybe optimize the branch predictor with https://stackoverflow.com/questions/109710/how-do-the-likely-unlikely-macros-in-the-linux-kernel-work-and-what-is-their-ben